src/test/scala/org/spartanz/parserz/prototype/T2.scala

```diff
@@ -0,0 +1,131 @@
+package org.spartanz.parserz.prototype
+
+import org.spartanz.parserz.ParsersModule
+
+object T2 {
+
+  object Module extends ParsersModule {
+    override type Input = List[Char]
+  }
+
+  import Module.Grammar._
+  import Module._
+  import Tools._
+
+  type S    = String
+  type E    = String
+  type G[A] = Grammar[S, S, E, A]
+
+  val one: G[Int]     = "one" @@ consumePure(i => i.tail -> i.head.toString.toInt, { case (i, a) => i ::: a.toString.toList })
+  val two: G[Boolean] = "two" @@ consumePure(i => i.tail -> (i.head == 'T'), { case (i, a) => i ::: (if (a) List('T') else List('F')) })
+  val thr: G[String]  = "thr" @@ succeed("blah")
+
+  case class Thing(idx: Int, exists: Boolean, name: String)
+
+  implicit val thingEquiv: Equiv[((Int, Boolean), String), Thing] =
+    Equiv.caseClass3((a, b, c) => Thing.apply(a, b, c), t => Thing.unapply(t))
+
+  val g1: G[((Int, Boolean), String)] = one ~ two ~ thr
+
+  val g2: G[Thing] = toZDirect(g1)
+  val g3: G[Thing] = toZ(g1)
+
+
+  object Tools {
+    import Hacks._
+
+    sealed trait Equiv[A, Repr]
+    case class Eq3[Z, A, B, C](f: (A, B, C) => Z, g: Z => Option[(A, B, C)]) extends Equiv[((A, B), C), Z]
+
+    object Equiv {
+//      def caseClass1[Z, A](f: A => Z, g: Z => Option[A]): Equiv[A, Z] = ???
+//      def caseClass2[Z, A, B](f: (A, B) => Z, g: Z => Option[(A, B)]): Equiv[(A, B), Z] = ???
+      def caseClass3[Z, A, B, C](f: (A, B, C) => Z, g: Z => Option[(A, B, C)]): Equiv[((A, B), C), Z] = Eq3(f, g)
+    }
```

Comment:
`Equiv` is used to enforce type safety for conversion between nested Tuple types to Case Class types